Create examples of working with Prefect REST APIs (#9661)
PrefectHQ/prefect: docs/api-ref/rest-api/index.md

tags:
    - REST API
    - Prefect Cloud
    - Prefect server
    - curl
    - PrefectClient
    - Requests
---

# REST API

The [Prefect REST API](https://deploy-preview-9661--prefect-docs-preview.netlify.app/api-ref/rest-api/) is used for communicating data from clients to the Prefect server so that orchestration can be performed. This API is consumed by clients such as the Prefect Python SDK or the server dashboard.

Prefect Cloud and a locally hosted Prefect server each provide a REST API.

- Prefect Cloud:

>     - Interactive Prefect Cloud REST API documentation is available at https://app.prefect.cloud/api/docs.

Reviewer: ```suggestion
    - Interactive Prefect Cloud REST API documentation is available at [https://app.prefect.cloud/api/docs](https://app.prefect.cloud/api/docs).
```

tags:
    - REST API
    - Prefect Cloud
    - Prefect server
    - curl
    - PrefectClient

>     - Requests

Reviewer: maybe add `API reference`?

---

> # REST API

Reviewer: > The [Prefect REST API](/api-ref/rest-api/)

@discdiver Not your fault but I'm noticing this is a self-link.

    - REST API
    - Prefect Cloud
    - Prefect Server

>     - curl

Reviewer: maybe `requests`?

# REST API

> The [Prefect REST API](https://deploy-preview-9661--prefect-docs-preview.netlify.app/api-ref/rest-api/) is used for communicating data from clients to the Prefect server so that orchestration can be performed. This API is consumed by clients such as the Prefect Python SDK or the server dashboard.

Reviewer: ```suggestion
The [Prefect REST API](/api-ref/rest-api/) is used for communicating data from clients to the Prefect server so that orchestration can be performed. This API is consumed by clients such as the Prefect Python SDK or the server dashboard.
```

Reviewer: why did they all get changed to deploy preview?

Author: I believe I copy-pasted your suggested change. Will fix those. Good catch!

Prefect Cloud and a locally hosted Prefect server each provide a REST API.

- Prefect Cloud:
    - Interactive Prefect Cloud REST API documentation is available at https://app.prefect.cloud/api/docs.
    - [Finding your Prefect Cloud details](#finding-your-prefect-cloud-details)
- Locally hosted open-source Prefect server:

>     - Interactive REST API documentation for a locally hosted open-source Prefect server is available at http://localhost:4200/docs or the /docs endpoint of the [PREFECT_API_URL](https://deploy-preview-9661--prefect-docs-preview.netlify.app/concepts/settings/#prefect_api_url) you have configured to access the server. You must have the server running with prefect server start to access the interactive documentation.

Reviewer: `prefect server start`

Reviewer: I think it would be easier to grab the actual markdown instead of trying to copy and paste in the UI because the links have changed and the code snippets too

> Prefect Cloud and a locally hosted Prefect server each provide a REST API.

Reviewer: I'm wondering if this sentence is still necessary now

Reviewer: 👍

- Prefect Cloud:

>     - [Interactive Prefect Cloud REST API documentation](https://app.prefect.cloud/api/docs).

Reviewer: ```suggestion
    - [Interactive Prefect Cloud REST API documentation](https://app.prefect.cloud/api/docs)
```

    - [Finding your Prefect Cloud details](#finding-your-prefect-cloud-details)
- Locally hosted open-source Prefect server:

>     - Interactive REST API documentation for a locally hosted open-source Prefect server is available at http://localhost:4200/docs or the `/docs` endpoint of the [PREFECT_API_URL](/concepts/settings/#prefect_api_url) you have configured to access the server. You must have the server running with `prefect server start` to access the interactive documentation.

Reviewer: ```suggestion
    - Interactive REST API documentation for a locally hosted open-source Prefect server is available at `http://localhost:4200/docs` or the `/docs` endpoint of the [PREFECT_API_URL](/concepts/settings/#prefect_api_url) you have configured to access the server. You must have the server running with `prefect server start` to access the interactive documentation.
```

it might make sense to at least add dashes so it's not just a bare link

Reviewer: similar to here:
https://deploy-preview-9661--prefect-docs-preview.netlify.app/concepts/infrastructure/?h=localhost#dockercontainer

Reviewer: Also maybe this documentation could be the second bullet point

>     - [Prefect REST API documentation](/api-ref/rest-api-reference/).

Reviewer: ```suggestion
    - [Prefect REST API documentation](/api-ref/rest-api-reference/)
```

        "key": {
            "exists_": True
        }
    }
}

headers = {"Authorization": f"Bearer {PREFECT_API_KEY}"}
endpoint = f"{PREFECT_API_URL}/artifacts/filter"

response = requests.post(endpoint, headers=headers, json=data)
assert response.status_code == 200
for artifact in response.json():
    print(artifact)
```

### curl with Prefect Cloud

This example uses curl with Prefect Cloud to create a flow run:

```bash
ACCOUNT_ID="abc-my-cloud-account-id-goes-here"
WORKSPACE_ID="123-my-workspace-id-goes-here"
PREFECT_API_URL="https://api.prefect.cloud/api/accounts/$ACCOUNT_ID/workspaces/$WORKSPACE_ID"
PREFECT_API_KEY="123abc_my_api_key_goes_here"
DEPLOYMENT_ID="my_deployment_id"

curl --location --request POST "$PREFECT_API_URL/deployments/$DEPLOYMENT_ID/create_flow_run" \
  --header "Content-Type: application/json" \
  --header "Authorization: Bearer $PREFECT_API_KEY" \
  --header "X-PREFECT-API-VERSION: 0.8.4" \
  --data-raw "{}"
```

Note that in this example `--data-raw "{}"` is required and is where you can specify other aspects of the flow run such as the state. Windows users substitute `^` for `\` for line multi-line commands.

## Finding your Prefect Cloud details

> When working with the Prefect Cloud REST API you will need your Account ID and often need the Workspace ID for the [workspace](/cloud/workspaces/) you want to interact with. You can find both IDs for a [Prefect profile](/concepts/settings/) in the CLI with `prefect profile inspect my_profile`. This command will also display your [Prefect API key](/cloud/users/api-keys/), as shown below:

Reviewer: ```suggestion
When working with the Prefect Cloud REST API you will need your Account ID and often the Workspace ID for the [workspace](/cloud/workspaces/) you want to interact with. You can find both IDs for a [Prefect profile](/concepts/settings/) in the CLI with `prefect profile inspect my_profile`. This command will also display your [Prefect API key](/cloud/users/api-keys/), as shown below:
```

- Interactive REST API documentation for a locally hosted open-source Prefect server is available at <a href="http://localhost:4200/docs" target="_blank">http://localhost:4200/docs</a> or the `/docs` endpoint of the [`PREFECT_API_URL`](/concepts/settings/#prefect_api_url) you have configured to access the server. You must have the server running with `prefect server start` to access the interactive documentation.
- The REST API documentation for a locally hosted open-source Prefect server is also available in the [Prefect REST API Reference](/api-ref/rest-api-reference/).

## Interacting with the REST API

> You have many options to interact with Prefect REST API

Author: ```suggestion
You have many options to interact with the Prefect REST API
```

> You have many options to interact with Prefect REST API
> - create an instance of [`PrefectClient`](/api-ref/prefect/client/orchestration/#prefect.client.orchestration.PrefectClient) 

Reviewer 2: ```suggestion
You have many options to interact with Prefect REST API:

- create an instance of [`PrefectClient`](/api-ref/prefect/client/orchestration/#prefect.client.orchestration.PrefectClient) 
```

Need a newline for bullets to render

> - use your favorite Python HTTP library such as [requests](https://requests.readthedocs.io/en/latest/) or [httpx](https://www.python-httpx.org/)

Reviewer 2: We should capitalize these to match our style guide

Reviewer: I read this as capitalize the first word of each bullet point, not capitalize `requests` 🤷‍♀️

> - use your favorite Python HTTP library such as [requests](https://requests.readthedocs.io/en/latest/) or [httpx](https://www.python-httpx.org/)

Author: ```suggestion
- use your favorite Python HTTP library such as [Requests](https://requests.readthedocs.io/en/latest/) or [HTTPX](https://www.python-httpx.org/)
```

- use an HTTP library in your language of choice

> - use [CURL](https://curl.se/) from the command line 

Reviewer: This isn't showing up as bullet points in the preview. Not sure if broken preview:
https://deploy-preview-9661--prefect-docs-preview.netlify.app/api-ref/rest-api/#interacting-with-the-rest-api

Here's an example of using `PrefectClient` with a locally hosted Prefect server:

```python
import asyncio
from prefect.client import get_client

async def get_flows():
    client = get_client()
    r = await client.read_flows(limit=5)
    return r

r = asyncio.run(get_flows())

for flow in r:
    print(flow.name, flow.id)

if __name__ == "__main__":
    asyncio.run(get_flows())
```

Output:

<div class="terminal">
```bash
cat-facts 58ed68b1-0201-4f37-adef-0ea24bd2a022
dog-facts e7c0403d-44e7-45cf-a6c8-79117b7f3766
sloth-facts 771c0574-f5bf-4f59-a69d-3be3e061a62d
capybara-facts fbadaf8b-584f-48b9-b092-07d351edd424
lemur-facts 53f710e7-3b0f-4b2f-ab6b-44934111818c
```
</div>

Here's an example of using CURL to create a flow run with Prefect Cloud:

> ```bash

Reviewer 2: Should probably hint to use `prefect config view` after logging in to find your API URL

ACCOUNT="my_cloud_account_can_be_found_in_my_prefect_cloud_url"

> WORKSPACE="my_cloud_workspace_can_be_found_in_my_prefect_cloud_url"

Reviewer 2: Might want to demo that these are UUIDs then include how to find it outside of the code blurb or with a comment

PREFECT_API_URL="https://api.prefect.cloud/api/accounts/$ACCOUNT/workspaces/$WORKSPACE"
PREFECT_API_KEY="my_api_key_goes_here"

> DEPLOYMENT="my-deployment"

Reviewer 2: This might need to be a UUID?

Author: It needed a few fixes - changes made.

- Interactive REST API documentation for a locally hosted open-source Prefect server is available at <a href="http://localhost:4200/docs" target="_blank">http://localhost:4200/docs</a> or the `/docs` endpoint of the [`PREFECT_API_URL`](/concepts/settings/#prefect_api_url) you have configured to access the server. You must have the server running with `prefect server start` to access the interactive documentation.
- The REST API documentation for a locally hosted open-source Prefect server is also available in the [Prefect REST API Reference](/api-ref/rest-api-reference/).

## Interacting with the REST API

You have many options to interact with the Prefect REST API:

- create an instance of [`PrefectClient`](/api-ref/prefect/client/orchestration/#prefect.client.orchestration.PrefectClient) 
- use your favorite Python HTTP library such as [Requests](https://requests.readthedocs.io/en/latest/) or [HTTPX](https://www.python-httpx.org/)
- use an HTTP library in your language of choice

> - use [curl]](https://curl.se/) from the command line 

Reviewer: These should be capitalized like Zanie mentioned

> - use [curl]](https://curl.se/) from the command line 

Reviewer: ```suggestion
- use [curl](https://curl.se/) from the command line 
```

### PrefectClient with Prefect server

> Here's an example that uses `PrefectClient` with a locally hosted Prefect server:

Reviewer 3: ```suggestion
This example uses `PrefectClient` with a locally hosted Prefect server:
```

```python
import asyncio
from prefect.client import get_client

async def get_flows():
    client = get_client()
    r = await client.read_flows(limit=5)
    return r

r = asyncio.run(get_flows())

for flow in r:
    print(flow.name, flow.id)

if __name__ == "__main__":
    asyncio.run(get_flows())
```

Output:

<div class="terminal">
```bash
cat-facts 58ed68b1-0201-4f37-adef-0ea24bd2a022
dog-facts e7c0403d-44e7-45cf-a6c8-79117b7f3766
sloth-facts 771c0574-f5bf-4f59-a69d-3be3e061a62d
capybara-facts fbadaf8b-584f-48b9-b092-07d351edd424
lemur-facts 53f710e7-3b0f-4b2f-ab6b-44934111818c
```
</div>

### Requests with Prefect

> Here's an example that uses the Requests library with Prefect Cloud to return the five newest artifacts.

Reviewer 3: ```suggestion
This example uses the Requests library with Prefect Cloud to return the five newest artifacts.
```

```python
import requests

PREFECT_API_URL="https://api.prefect.cloud/api/accounts/abc-my-cloud-account-id-is-here/workspaces/123-my-workspace-id-is-here"
PREFECT_API_KEY="123abc_my_api_key_goes_here"
data = {
    "sort": "CREATED_DESC",
    "limit": 5,
    "artifacts": {
        "key": {
            "exists_": True
        }
    }
}

headers = {"Authorization": f"Bearer {PREFECT_API_KEY}"}
endpoint = f"{PREFECT_API_URL}/artifacts/filter"

response = requests.post(endpoint, headers=headers, json=data)
assert response.status_code == 200
for artifact in response.json():
    print(artifact)
```

### curl with Prefect Cloud

> Here's an example that uses curl with Prefect Cloud to create a flow run:

Reviewer 3: ```suggestion
This example uses curl with Prefect Cloud to create a flow run:
```

```bash
ACCOUNT_ID="abc-my-cloud-account-id-goes-here"
WORKSPACE_ID="123-my-workspace-id-goes-here"

> PREFECT_API_URL="https://api.prefect.cloud/api/accounts/$ACCOUNT_ID/workspaces/$WORKSPACE_ID""

Reviewer: ```suggestion
PREFECT_API_URL="https://api.prefect.cloud/api/accounts/$ACCOUNT_ID/workspaces/$WORKSPACE_ID"
```

PREFECT_API_KEY="123abc_my_api_key_goes_here"
DEPLOYMENT_ID="my_deployment_id"

curl --location --request POST "$PREFECT_API_URL/deployments/$DEPLOYMENT_ID/create_flow_run" \
  --header "Content-Type: application/json" \
  --header "Authorization: Bearer $PREFECT_API_KEY" \
  --header "X-PREFECT-API-VERSION: 0.8.4" \
  --data-raw "{}"
```

Note that in this example `--data-raw "{}"` is required and is where you can specify other aspects of the flow run such as the state. Windows users substitute `^` for `\` for line multi-line commands.

## Finding your Prefect Cloud details

When working with the Prefect Cloud REST API you will need your Account ID and often need the Workspace ID for the [workspace](/cloud/workspaces/) you want to interact with. You can find both IDs for a [Prefect profile](/concepts/settings/) in the CLI with `prefect profile inspect my_profile`. This command will also display your [Prefect API key](/cloud/users/api-keys/), as shown below:

<div class="terminal">
```bash
PREFECT_API_URL='https://api.prefect.cloud/api/accounts/abc-my-account-id-is-here/workspaces/123-my-workspace-id-is-here'
PREFECT_API_KEY='123abc_my_api_key_is_here'
```
</div>

> Alternatively, view your Account ID and Workspace ID in your browser URL. For example: `https://app.prefect.cloud/account/abc-my-account-id-is-here/workspaces/123-my-workspace-id-is-here`. 

Reviewer: It would be nice to know this ahead of time

> Alternatively, view your Account ID and Workspace ID in your browser URL. For example: `https://app.prefect.cloud/account/abc-my-account-id-is-here/workspaces/123-my-workspace-id-is-here`. 

Author: Think I should just move that whole section on finding your cloud details above the examples of interacting with the API, @serinamarie ?

> Alternatively, view your Account ID and Workspace ID in your browser URL. For example: `https://app.prefect.cloud/account/abc-my-account-id-is-here/workspaces/123-my-workspace-id-is-here`. 

Reviewer: I would consider something like:

The [Prefect REST API](https://deploy-preview-9661--prefect-docs-preview.netlify.app/api-ref/rest-api/) is used for communicating data from clients to the Prefect server so that orchestration can be performed. This API is consumed by clients such as the Prefect Python SDK or the server dashboard.

Prefect Cloud and a locally hosted Prefect server each provide a REST API.
- Prefect Cloud:
    - Interactive Prefect Cloud REST API documentation is available at https://app.prefect.cloud/api/docs.
    - Finding your Prefect Cloud details
- Locally hosted open-source Prefect server:
    - Interactive REST API documentation for a locally hosted open-source Prefect server is available at http://localhost:4200/docs or the /docs endpoint of the [PREFECT_API_URL](https://deploy-preview-9661--prefect-docs-preview.netlify.app/concepts/settings/#prefect_api_url) you have configured to access the server. You must have the server running with prefect server start to access the interactive documentation.
    - The REST API documentation for a locally hosted open-source Prefect server is also available in the [Prefect REST API Reference](https://deploy-preview-9661--prefect-docs-preview.netlify.app/api-ref/rest-api-reference/).

> Alternatively, view your Account ID and Workspace ID in your browser URL. For example: `https://app.prefect.cloud/account/abc-my-account-id-is-here/workspaces/123-my-workspace-id-is-here`. 

Author: done

- Interactive REST API documentation for a locally hosted open-source Prefect server is available at <a href="http://localhost:4200/docs" target="_blank">http://localhost:4200/docs</a> or the `/docs` endpoint of the [`PREFECT_API_URL`](/concepts/settings/#prefect_api_url) you have configured to access the server. You must have the server running with `prefect server start` to access the interactive documentation.
- The REST API documentation for a locally hosted open-source Prefect server is also available in the [Prefect REST API Reference](/api-ref/rest-api-reference/).

## Interacting with the REST API

You have many options to interact with the Prefect REST API:

- Create an instance of [`PrefectClient`](/api-ref/prefect/client/orchestration/#prefect.client.orchestration.PrefectClient) 
- Use your favorite Python HTTP library such as [Requests](https://requests.readthedocs.io/en/latest/) or [HTTPX](https://www.python-httpx.org/)
- Use an HTTP library in your language of choice
- Use [curl](https://curl.se/) from the command line 

> ### PrefectClient with Prefect server

Reviewer: ```suggestion
### `PrefectClient` with Prefect server
```

Reviewer: Not sure of the docs style guide on headings but in the preview it looks strange to me, specifically "curl with Prefect" without the code dashes on `curl`

Author: curl's website rolls similarly, so I think it's ok as is. :)
